Utilities/docker

```diff
@@ -0,0 +1,163 @@
+
+ Note that currently this script is also responsible for building the package
+ manager in such a way that it can be installed along with the Swift package.
+ In particular, it knows how to build the runtime PackageDescription library
+ correctly and install it.  It can also build libSwiftPM, allowing clients to
+ access package manager functionality.
+"""
+
+import argparse
+import urllib2
+import os
+import subprocess
+import yaml
+
+def call(args):
+    """Prints and executes a command."""
+    print " ".join(args)
+    return subprocess.call(args)
+
+def docker_run(command_args):
+    """Runs a command in the container."""
+    call([
+        "docker",
+        "run",
+        "-it",
+        "--security-opt", "seccomp=unconfined",
+        "-v", os.getcwd() + ":/swiftpm",
+        "-w", "/swiftpm",
+        "--rm",
+        "swiftpm-docker-1604"
+    ] + command_args)
+
+def build(_):
+    """Builds a docker image with the latest snapshot."""
+
+    # Get latest snapshot info.
+    latest_build_url = "https://swift.org/builds/development/ubuntu1604/latest-build.yml"
+    latest_build_data = yaml.load(urllib2.urlopen(latest_build_url).read())
+    snapshot_filename = latest_build_data["download"]
+    # FIXME: We shouldn"t need to do this, it should be available in the API.
+    snapshot_name = snapshot_filename.replace("-ubuntu16.04.tar.gz", "")
+    base_url = latest_build_url.rsplit("/", 1)[0]
+    latest_snapshot_url = base_url + "/" + snapshot_name + "/" + snapshot_filename
+    script_dir = os.path.dirname(os.path.realpath(__file__))
+
+    # Download latest snapshot (if necessary).
+    if snapshot_filename in os.listdir(script_dir):
+        print "Not downloading: " + latest_snapshot_url
+    else:
+        result = call([
+            "curl",
+            "-o",
+            script_dir + "/" + snapshot_filename,
+            latest_snapshot_url
+        ])
+
+        if result != 0:
+            print "Unable to install package: " + latest_snapshot_url
+            exit(1)
+
+    # Create docker image.
+    call([
+        "docker",
+        "build",
+        "-t", "swiftpm-docker-1604",
+        "--build-arg", "SNAPSHOT=%s" % snapshot_filename,
+        script_dir
+    ])
+
+def run(args):
+    """Runs a command in the container."""
+    docker_run([args.command] + args.arguments)
+
+def bootstrap(args):
+    """Runs the bootstrap script in the container."""
+    docker_run(["Utilities/bootstrap"] + args.arguments)
+
+def swift_build(args):
+    """Runs the swift-build executable in the container."""
+    docker_run([".build/x86_64-unknown-linux/debug/swift-build"] + args.arguments)
+
+def swift_test(args):
+    """Runs the swift-test executable in the container."""
+    docker_run([".build/x86_64-unknown-linux/debug/swift-test"] + args.arguments)
+
+def swift_run(args):
+    """Runs the swift-run executable in the container."""
+    docker_run([".build/x86_64-unknown-linux/debug/swift-run"] + args.arguments)
```

Comment:
We can unify these into one function, right?

Reply:
You mean something like this?

```python
def run_swift_executable(executable, args):
    """Runs an executable from the build folder in the container."""
    docker_run([".build/x86_64-unknown-linux/debug/" + executable] + args.arguments)

def swift_run(command, args):
    """Runs the swift-run executable in the container."""
    run_swift_executable("swift-run", args)
```

Reply:
Yeah but we don't even need the latter 2 functions.

Reply:
I kept those so that main doesn't need to do any branching on the subcommand:

```python
args = parser.parse_args()
args.func(args)
```

Reply:
Can we use the argument to infer what was passed?

Reply:
Yep, much cleaner. Done.